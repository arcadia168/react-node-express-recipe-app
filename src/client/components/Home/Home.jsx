import React, { Component } from 'react';
import { ListGroup, ListGroupItem } from 'react-bootstrap';
import '../../scss/application.scss';

class Home extends Component {
    constructor(props) {
        super(props);
        this.state = {
            recipes: [],
        };
    }

    componentDidMount() {
        //then make a call for recipes
        this.props.recipeService.getRecipes(false).then((recipes) => {
            this.setState({
                recipes: recipes
            })
        });
    }

    render() {
        const recipeList = this.state.recipes;

        if (recipeList) {
            return <ListGroup>Hello World, here are my recipes: {recipeList.map((recipe, index, recipes) => {
                return <ListGroupItem href={`/recipe/${recipe._id}`} header={recipe.name} key={index}>{recipe.cookingTime}{recipe.mainIngredients.join(', ')}</ListGroupItem>
            })} 
            </ListGroup>;
        }
        else {
            return <div>Nothing to see here in HOME!</div>
        }
    }
}
module.exports = Home;