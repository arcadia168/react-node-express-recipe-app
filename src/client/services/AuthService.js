import auth0 from 'auth0-js';
import history from './HistoryService'
import axios from 'axios';

class AuthService {
    constructor() {
        this.auth0 = new auth0.WebAuth({
            domain: 'bennawazcodedemos.eu.auth0.com',
            clientID: 'B5bvppKVgfaM8cj3eukBOhoShRq0eBBE',
            redirectUri: 'http://localhost:3000/callback',
            audience: 'react-node-recipes-auth-api',
            responseType: 'token id_token',
            scope: 'openid profile email'
        });

        this.userProfile = undefined;
        this.login = this.login.bind(this);
        this.logout = this.logout.bind(this);
        this.handleAuthentication = this.handleAuthentication.bind(this);
        this.isAuthenticated = this.isAuthenticated.bind(this);
    }

    handleAuthentication() {
        this.auth0.parseHash((err, authResult) => {
            if (authResult && authResult.accessToken && authResult.idToken) {
                this.setSession(authResult);
                history.replace('/home');
            } else if (err) {
                history.replace('/home');
                console.log(err);
            }
        });
    }

    getUserProfile() {
        if (!this.userProfile) {
            var accessToken = localStorage.getItem('access_token');

            if (!accessToken) {
                console.log('Access token must exist to fetch profile');
            }

            this.auth0.client.userInfo(accessToken, (err, profile) => {
                if (profile) {
                    this.userProfile = profile;
                }
            });
        } else {
            return this.userProfile;
        }
    }

    updateUserProfile(user_profile) {
        axios.post('/api/users', user_profile).then((updatedUserProfile) => {
            //use the favourite recipes
            console.log('User profile updated on backend API: ' + JSON.stringify(updatedUserProfile));
            localStorage.setItem('user_profile', updatedUserProfile);
        }).catch((error) => {
            //log error
            console.log(error);
        });
    }

    setSession(authResult) {
        // Set the time that the access token will expire at
        let expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
        localStorage.setItem('access_token', authResult.accessToken);
        localStorage.setItem('id_token', authResult.idToken);
        localStorage.setItem('expires_at', expiresAt);
        localStorage.setItem('user_profile', authResult.idTokenPayload);
        //set the axios service to use this auth header
        ;
        axios.defaults.headers.common['Authorization'] = `Bearer ${localStorage.getItem('access_token')}`;
        //send the user profile to the backend and store/update it in the db
        this.updateUserProfile(authResult.idTokenPayload);
        // navigate to the home route
        history.replace('/home');
    }

    logout() {
        // Clear access token and ID token from local storage
        localStorage.removeItem('access_token');
        localStorage.removeItem('id_token');
        localStorage.removeItem('expires_at');
        // navigate to the home route
        history.replace('/home');
    }

    isAuthenticated() {
        // Check whether the current time is past the 
        // access token's expiry time
        let expiresAt = JSON.parse(localStorage.getItem('expires_at'));
        return new Date().getTime() < expiresAt;
    }

    login() {
        this.auth0.authorize();
    }
}

module.exports = AuthService;