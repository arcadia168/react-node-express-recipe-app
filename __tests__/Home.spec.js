import React from 'react';
import ReactDOM from 'react-dom';
import Home from 'Home';
import HistoryService from 'HistoryService';
import AuthService from 'AuthService';
import RecipeService from 'RecipeService';
import axios from 'axios';
import renderer from 'react-test-renderer';
const axiosInstance = axios;
const recipeServiceInstance = new RecipeService(axiosInstance);
const authInstance = new AuthService(axiosInstance);

describe('Home', () => {
    it('renders correctly', () => {
        const tree = renderer
          .create(<Home axios={axiosInstance} recipeService={recipeServiceInstance} auth={authInstance}/>)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
});