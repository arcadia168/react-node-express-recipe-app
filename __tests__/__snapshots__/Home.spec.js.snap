// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<div>
  Nothing to see here in RECIPE!
</div>
`;
